components/zcash_address/src/kind/unified/fvk.rs

```diff
@@ -0,0 +1,360 @@
+            Typecode::P2sh => Err(data),
+            Typecode::Sapling => data.try_into().map(Fvk::Sapling),
+            Typecode::Orchard => data.try_into().map(Fvk::Orchard),
+            Typecode::Unknown(_) => Ok(Fvk::Unknown { typecode, data }),
+        }
+        .map_err(|e| {
+            ParseError::InvalidEncoding(format!("Invalid fvk for typecode {}: {:?}", typecode, e))
+        })
+    }
+}
+
+impl SealedItem for Fvk {
+    fn typecode(&self) -> Typecode {
+        match self {
+            Fvk::P2pkh(_) => Typecode::P2pkh,
+            Fvk::Sapling(_) => Typecode::Sapling,
+            Fvk::Orchard(_) => Typecode::Orchard,
+            Fvk::Unknown { typecode, .. } => Typecode::Unknown(*typecode),
+        }
+    }
+
+    fn data(&self) -> &[u8] {
+        match self {
+            Fvk::P2pkh(data) => data,
+            Fvk::Sapling(data) => data,
+            Fvk::Orchard(data) => data,
+            Fvk::Unknown { data, .. } => data,
+        }
+    }
+}
+
+/// A Unified Full Viewing Key.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Ufvk(pub(crate) Vec<Fvk>);
+
+impl Container for Ufvk {
+    type Item = Fvk;
+
+    /// Returns the FVKs contained within this UFVK, in the order they were
+    /// parsed from the string encoding.
+    ///
+    /// This API is for advanced usage; in most cases you should use `Ufvk::receivers`.
+    fn items_as_parsed(&self) -> &[Fvk] {
+        &self.0
+    }
+}
+
+impl Encoding for Ufvk {}
+
+impl SealedContainer for Ufvk {
+    /// The HRP for a Bech32m-encoded mainnet Unified FVK.
+    ///
+    /// Defined in [ZIP 316][zip-0316].
+    ///
+    /// [zip-0316]: https://zips.z.cash/zip-0316
+    const MAINNET: &'static str = "uview";
+
+    /// The HRP for a Bech32m-encoded testnet Unified FVK.
+    ///
+    /// Defined in [ZIP 316][zip-0316].
+    ///
+    /// [zip-0316]: https://zips.z.cash/zip-0316
+    const TESTNET: &'static str = "uviewtest";
+
+    /// The HRP for a Bech32m-encoded regtest Unified FVK.
+    const REGTEST: &'static str = "uviewregtest";
+
+    fn from_inner(fvks: Vec<Self::Item>) -> Self {
+        Self(fvks)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use assert_matches::assert_matches;
+
+    use proptest::{array::uniform32, prelude::*, sample::select};
+
+    use super::{Fvk, ParseError, Typecode, Ufvk};
+    use crate::{
+        kind::unified::{private::SealedContainer, Container, Encoding},
+        Network,
+    };
+
+    prop_compose! {
+        fn uniform128()(a in uniform32(0u8..), b in uniform32(0u8..), c in uniform32(0u8..), d in uniform32(0u8..)) -> [u8; 128] {
+            let mut fvk = [0; 128];
+            fvk[..32].copy_from_slice(&a);
+            fvk[32..64].copy_from_slice(&b);
+            fvk[64..96].copy_from_slice(&c);
+            fvk[96..].copy_from_slice(&d);
+            fvk
+        }
+    }
+
+    prop_compose! {
+        fn uniform96()(a in uniform128()) -> [u8; 96] {
```

Comment:
How does proptest minimisation behave when some of the generated input goes completely unused? Should these APIs actually be constructed the same way (duplicating code), or in reverse (smaller to bigger)?

Reply:
Hm, this is a good point; smaller to bigger is probably the correct way.